### PR TITLE
SGD optimization & added test

### PR DESCRIPTION
**Pull request by ap-hynninen**
Optimizes SGD by fusing multiple kernel calls into a single Kernel::Launch -kernel

### CODE REVIEW

File: src/operator/operator_common.h [hunk 1]
```diff
@@ -43,7 +59,7 @@ namespace op {
         LOG(FATAL) << "not reached";    \
```
Comment (piiswrong) [inline src/operator/operator_common.h]: we can remove this for both cpu and gpu

Comment (piiswrong) [inline src/operator/operator_common.h]: actually no. we have a Kernel_assign for kernels. The assign is for host code

File: src/operator/optimizer_op-inl.h [hunk 2]
```diff
@@ -42,32 +43,42 @@ struct SGDParam : public dmlc::Parameter<SGDParam> {
   }
 };
 
+struct SGDKernel {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* out_data, const DType* weight_data,
+    const DType* grad_data, const DType param_clip_gradient,
+    const DType param_lr, const DType param_wd, const DType param_rescale_grad,
+    const OpReqType req) {
+    if (param_clip_gradient >= 0.0f) {
```
Comment (piiswrong) [inline src/operator/optimizer_op-inl.h]: would it be faster if you do this in host code?

Comment (ap-hynninen) [inline src/operator/optimizer_op-inl.h]: I tested it, does not affect performance much

**Pull request merged**